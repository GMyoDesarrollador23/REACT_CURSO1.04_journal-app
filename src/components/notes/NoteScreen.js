import Swal from "sweetalert2";
import React, { useEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";

import { NoteAppBar } from "./NoteAppBar";
import { useForm } from "../../hooks/useForm";
import {
   activeNote,
   startDeleting,
} from "../../actions/notes";

export const NoteScreen = () => {
   const dispatch = useDispatch();

   const { active: note } = useSelector(
      (state) => state.notes
   );

   const [formValues, handleInputChange, reset] = useForm(
      note
   );
   const { title, body, id } = formValues;

   const activeId = useRef(note.id);

   useEffect(() => {
      if (note.id !== activeId.current) {
         reset(note);
         activeId.current = note.id;
      }
   }, [note, reset]);

   useEffect(() => {
      dispatch(
         activeNote(formValues.id, { ...formValues })
      );
   }, [formValues, dispatch]);

   const handleDelete = () => {
      Swal.fire({
         title: "Delete Note",
         text: "Are you sure you want to delete this note?",
         icon: "warning",
         showCancelButton: true,
         confirmButtonColor: "#3085d6",
         cancelButtonColor: "#d33",
         confirmButtonText: "Yes, delete it!",
      }).then((result) => {
         if (result.value) {
            dispatch(startDeleting(id));
            Swal.fire({
               position: "center",
               icon: "success",
               title: "note cleared with success",
               showConfirmButton: false,
               timer: 1500,
            });
         }
      });
   };

   return (
      <div className="notes__main-content animate__animated animate__fadeIn animate__faster">
         <NoteAppBar />

         <div className="notes__content">
            <input
               type="text"
               placeholder="hola"
               className="notes__title-input"
               autoComplete="off"
               name="title"
               value={title}
               onChange={handleInputChange}
            />
            <textarea
               placeholder="que paso ayer"
               className="notes__texaria"
               name="body"
               value={body}
               onChange={handleInputChange}
            ></textarea>
            {note.url && (
               <div className="notes__image animate__animated animate__fadeIn animate__faster">
                  <img src={note.url} alt="paisaje" />
               </div>
            )}
         </div>
         <button
            className="btn btn-danger"
            onClick={handleDelete}
         >
            Delete
         </button>
      </div>
   );
};
