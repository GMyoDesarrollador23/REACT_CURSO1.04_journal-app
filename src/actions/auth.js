import Swal from "sweetalert2";

import { types } from "../types/types";
import {
   firebase,
   googleAuthProvider,
} from "../firebase/firebase-config";
import { finishLoading, startLoading } from "./ui";

import { notesLogout } from "../actions/notes";

export const startLoginEmailPassword = (
   email,
   password
) => {
   return (dispatch) => {
      dispatch(startLoading());
      return firebase
         .auth()
         .signInWithEmailAndPassword(email, password)
         .then(({ user }) => {
            dispatch(login(user.uid, user.displayName));

            dispatch(finishLoading());
            Swal.fire({
               position: "center",
               icon: "success",
               title: `welcome back "${user.displayName}"`,
               showConfirmButton: false,
               timer: 2000,
            });
         })
         .catch((e) => {
            console.warn(e);
            dispatch(finishLoading());
            Swal.fire({
               position: "center",
               icon: "error",
               title: e.message,
               showConfirmButton: false,
               timer: 2000,
            });
         });
   };
};

export const startRegisterWithEmailPasswordName = (
   email,
   password,
   name
) => {
   return (dispatch) => {
      firebase
         .auth()
         .createUserWithEmailAndPassword(email, password)
         .then(async ({ user }) => {
            await user.updateProfile({
               displayName: name,
            });
            // console.log(user);

            dispatch(login(user.uid, user.displayName));
         })
         .catch((e) => {
            console.log(e);
            Swal.fire({
               position: "center",
               icon: "error",
               title: e.message,
               showConfirmButton: false,
               timer: 2000,
            });
         });
   };
};

export const startGoogleLogin = () => {
   return (dispatch) => {
      firebase
         .auth()
         .signInWithPopup(googleAuthProvider)
         .then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
         });
   };
};

export const startLogoud = () => {
   return async (dispatch) => {
      await firebase.auth().signOut();

      dispatch(logout());
      dispatch(notesLogout());
   };
};

export const login = (uid, displayName) => ({
   type: types.login,
   payload: {
      uid,
      displayName,
   },
});

export const logout = () => ({
   type: types.logout,
});
