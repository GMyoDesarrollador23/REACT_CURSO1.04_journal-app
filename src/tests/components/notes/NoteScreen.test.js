import React from "react";
import { mount } from "enzyme";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";

import thunk from "redux-thunk";
import configureStore from "redux-mock-store";

import { activeNote } from "../../../actions/notes";
import { NoteScreen } from "../../../components/notes/NoteScreen";

// *****************************************************

jest.mock("../../../actions/notes", () => ({
   activeNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
   auth: {},
   ui: {
      loading: false,
      msgError: null,
   },
   notes: {
      active: {
         id: 1234,
         title: "hola",
         body: "Mundo",
         date: 0,
      },
      notes: [],
   },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wraper = mount(
   <Provider store={store}>
      <NoteScreen />
   </Provider>
);

// *****************************************************
describe("Pruebas en <NoteScreen/>", () => {
   // ---------------------------------------------
   test("debe de mostrarse correctamente", () => {
      expect(wraper).toMatchSnapshot();
   });

   // ---------------------------------------------
   test("debe de disparar el activeNote", () => {
      wraper
         .find('input[name="title"]')
         .simulate("change", {
            target: {
               name: "title",
               value: "texto de prueba",
            },
         });
      expect(activeNote).toHaveBeenLastCalledWith(1234, {
         title: "texto de prueba",
         body: "Mundo",
         id: 1234,
         date: 0,
      });
   });
   // ---------------------------------------------
});
