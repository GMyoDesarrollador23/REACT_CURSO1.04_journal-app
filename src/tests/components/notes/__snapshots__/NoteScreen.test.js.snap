// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen/> debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn animate__faster"
>
  <div
    className="notes__appbar"
  >
    <span>
      December 31st 1969, 8:00
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="hola"
      type="text"
      value="hola"
    />
    <textarea
      className="notes__texaria"
      name="body"
      onChange={[Function]}
      placeholder="que paso ayer"
      value="Mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
